readthedocs/proxito/views/serve.py

```diff
@@ -673,32 +673,45 @@ def _get_custom_404_page(self, request, project, version=None):
             if default_version:
                 versions_404.append(default_version)
 
+        if not versions_404:
+            return None
+
+        tryfiles = ["404.html", "404/index.html"]
+        available_404_files = list(
+            HTMLFile.objects.filter(
+                version__in=versions_404, path__in=tryfiles
+            ).values_list("version__slug", "path")
+        )
```

Comment:
Also, do we have an issue to track this work? If not, we should create one so we reduce the data we save in this table.

Reply:
```
In [22]: p = Project.objects.get(slug='docs')

In [23]: stable = p.versions.get(slug='stable')

In [24]: latest = p.versions.get(slug='latest')

In [25]: tryfiles = ['404.html', '404/index.html', 'intro/getting-started-with-mkdocs.html', 'intro/getting-started-foo/index.html']

In [26]: HTMLFile.objects.filter(version__in=[stable, latest], path__in=tryfiles).values_list('version__slug', 'path')

{'sql': 'SELECT "builds_version"."slug", "projects_importedfile"."path" FROM "projects_importedfile" INNER JOIN "builds
_version" ON ("projects_importedfile"."version_id" = "builds_version"."id") WHERE ("projects_importedfile"."name"::text
 LIKE \'%.html\' AND "projects_importedfile"."path" IN (\'404.html\', \'404/index.html\', \'intro/getting-started-with-
mkdocs.html\', \'intro/getting-started-foo/index.html\') AND "projects_importedfile"."version_id" IN (2604018, 2603893)
) LIMIT 21',
 'time': '0.005'}
```

This is a query with 4 files and two versions, it takes between 0.001 and 0.005. My only worry is that we will have an increase in queries to the DB, but probably isn't that bad, since these requests will be cached at the CDN level.

Reply:
>  Also, do we have an issue to track this work? If not, we should create one so we reduce the data we save in this table.

https://github.com/readthedocs/readthedocs.org/issues/10623

> If you are not going to implement this on this PR, please create an issue referencing to this one to track this work.

https://github.com/readthedocs/readthedocs.org/issues/10659

Reply:
Great, thanks!

Reply:
This query can be faster based on a DB cache if it was run multiple times, so we'll have to check how fast it averages our in production, but it'll definitely be faster than hitting storage!